src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1738,21 +1746,40 @@ def client_side_proxy_wrapper(cmd,
             raise FileOperationError("Failed to load credentials." + str(e))
 
         user_name = account['user']['name']
+        use_msal_cache = utils.use_msal_cache()
+        if not use_msal_cache:
+            if user_type == 'user':
+                key = 'userId'
+                key2 = 'refreshToken'
+            else:
+                key = 'servicePrincipalId'
+                key2 = 'accessToken'
 
-        if user_type == 'user':
-            key = 'userId'
-            key2 = 'refreshToken'
-        else:
-            key = 'servicePrincipalId'
-            key2 = 'accessToken'
-
-        for i in range(len(creds_list)):
-            creds_obj = creds_list[i]
-
-            if key in creds_obj and creds_obj[key] == user_name:
-                creds = creds_obj[key2]
-                break
+            for i in range(len(creds_list)):
+                creds_obj = creds_list[i]
 
+                if key in creds_obj and creds_obj[key] == user_name:
+                    creds = creds_obj[key2]
+                    break
+        else:
+            if user_type == "user":
+                response_user_objectid = az_cli("ad signed-in-user show --query objectId -o tsv")
+                token_cache_location = os.path.expanduser(os.path.join('~', msal_token_cache_user))
+                persistence = msal_extensions.FilePersistenceWithDataProtection(token_cache_location)
+                token_cache = msal_extensions.PersistedTokenCache(persistence)
+                token_cache._reload_if_necessary()
+                home_account_id = response_user_objectid + "." + tenantId
+                owned_by_home_account = {"home_account_id": home_account_id}
+                creds_info = token_cache.find(PersistedTokenCache.CredentialType.REFRESH_TOKEN, query=owned_by_home_account)
```

Comment:
@rayluo, I think this is another instance where people are hacking MSAL cache to get the refresh token.

Reply:
That is really unfortunate.

@sirireddy12 , why is refresh token (RT) needed? Could the usage pattern be replaced by periodically requesting an access token, possibly via Azure CLI? (@jiasli , does `az get-access-token ...` has an equivalent helper function to be called by extensions?)

FYI: Not only the token cache helpers are considered internal, MSAL would probably NOT save RT in token cache in near future.

Reply:
@rayluo  Refresh token is needed to retrieve PoP token. Does PoP implementation support available in MSAL python which az extension can utilize?

Reply:
> @rayluo Refresh token is needed to retrieve PoP token. Does PoP implementation support available in MSAL python which az extension can utilize?

MSAL Python does not currently support PoP. However, I got an impression that the PoP implementation might be similar to [another feature named "Ssh Cert" which MSAL Python does support](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.16.0/tests/test_e2e.py#L215-L238), and then there is also an az extension "ssh" already handles the key management and feeds signed content to MSAL. You may take a look and see whether that pattern could be repurposed for PoP token.